keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,308 @@
+    * Taint-based pod eviction latency 
+
+### Dependencies
+
+* **Does this feature depend on any specific services running in the cluster?** 
+    * N/A
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+    * No
+* **Will enabling / using this feature result in introducing new API types?** 
+    * No
+* **Will enabling / using this feature result in any new calls to the cloud provider?**
+    * No
+* **Will enabling / using this feature result in increasing size or count of the existing API objects?**
+    * No
+* **Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?**
+    * No
+* **Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?**
+    * No
+
+### Troubleshooting
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+    * The same behavior as the current version: fail to add taint and/or evict pods on tainted nodes 
+* **What are other known failure modes?**
+    * No
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+    * If the default or custom taint-manager is working or not. 
+
+## Implementation History
+
+* 2023-03-06: Initial KEP published. 
+
+## Drawbacks
+
+## Alternatives
+
+* **Leverage the existing tolerations for `NoExecute` taints**: operators opt-out from taint-based eviction by injecting tolerations for NoExecute taints. It has a few major disadvantages over an independent `TaintManager`.
+    * It will require to either inject tolerations to all pods via mutating webhooks and/or changing the manifests of all the running workloads. 
+    * It will require either injecting tolerations to every single pod via mutating webhooks and/or changing the manifests of all the running workloads.
+    * It will still need an additional controller to really evict after coordinating with the external reservation/lease system. This will add additional overhead to toleration injections.
+    * Custom mutating webhook may interfere with the admission plugin that applies the default tolerations if they are trying to apply tolerations for similar taints.
+    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.
+
+* **Enhance the existing toleration APIs**
+    * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establish whether eviction for a given pod is allowed or not.
```

Comment:
There are two APIs that allow extensibility of kubernetes without disabling default behavior:
- finalizers https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/
- scheduling gates https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/

Could you consider similar approaches?

I don't think the proposal in this paragraph is detailed enough, hence why it might not feel sufficient for your use case.

Reply:
> Could you consider similar approaches?


The only closest thing that I can think of is scheduler extender(similar to entity which is outside of Kube system).

>  hence why it might not feel sufficient for your use case.

Which use-case are you referring to here? Splitting taint-manager?

Reply:
Splitting the controllers is not a use-case. It's an implementation detail.

I'm not sure what use case the authors are trying to solve. I suppose it's close to story 1.

> The only closest thing that I can think of is scheduler extender

That goes along the way of "I want to fully override the component", which is the same as this proposal.

Reply:
Right, that could be a valid avenue towards providing flexibility. However, not all users have access to the control plane to configure the webhook. So a more generic API or something more similar to finalizers could be more usable.

Reply:
Agreed. I think the changes that @yuanchen8911 is suggesting are at the API level and he is also against having a webhook. However the implementation details may vary. Having said that, I guess we're debating about how we want to implement a gap in Kube in the alternate section of this KEP. The gap and how it needs to be addressed warrants another KEP. So, if you've any objections with the wording or the content(including the alternatives section), please let us know and @yuanchen8911 can make changes to move forward.

Reply:
I already expressed my point: I'm in favor of the refactoring. I disagree with the Motivation and Stories. Given the current motivations, I consider that the Alternatives section hand-waves that improvements to the API wouldn't work without going into much detail.

Reply:
I agree with your assessment on the motivation and use-cases section. 

However in the alternatives section, https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R301 the KEP states that the goal is not to change the API at this point of time. 

I believe https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R299 is relevant only to using webhooks not extending API. We feel that extending API is a good alternative but it is a larger change and it needs addressing few gaps that we mentioned. We will change the wording if that is what concerns you. Let me know if you have any other concerns.

Reply:
> I also disagree with that stance.

What would be your recommendation, Aldo?  Any specific things you want us to add to the KEP?

Reply:
I guess, the point of contention here is `the side-effect of splitting the controllers is that it allows cluster operators to build custom TaintManagers and disable the default TaintManager`. @yuanchen8911 went ahead and explained why it is beneficial to split from that standpoint. As I mentioned earlier, we cannot seriously explore alternatives in this KEP as we may digress from the original point. So, better to cut short and explain this KEP on separation of concerns and tell the side-effect of this split is people can disable default `TaintManager`.

Reply:
I like that approach better. It doesn't give false expectations to other people following the KEP.

Reply:
Updated the Non-Goals to clarify the code reorganizations the main goal of the KEP as below.

### Non-Goals

The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to the `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.

* Extend and enhance the current built-in APIs for `TaintManager` customization, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
* Actual extension and enhancement  of the current `TaintManager`.